File: Source/DoubleTagDeleter.cs
namespace NhlStenden.TagBag;

/// <summary>
/// Represents a class that deletes duplicate tags from files.
/// </summary>
public class DoubleTagDeleter : TagBase
{
    /// <summary>
    /// Executes the tag deletion process on the specified files.
    /// </summary>
    /// <param name="args">A list of arguments where the first argument is the directory path.</param>
    public static void Execute(IList<string> args)
    {
        var fileNames = GetTagFiles(args[0]);
        foreach (var fileName in fileNames)
        {
            ProcessFile(fileName);
        }
    }

    /// <summary>
    /// Processes the specified file to remove duplicate tags.
    /// </summary>
    /// <param name="fileName">The name of the file to process.</param>
    static void ProcessFile(string fileName)
    {
        var lines = File.ReadAllLines(fileName).ToList();

        for (var i = 0; i < lines.Count; i++)
        {
            var line = lines[i];
            var newLine = new List<string>();

            var items = TagHelper.SplitIntoTags(line);
            for (var j = 0; j < items.Count; j++)
            {
                var item = items[j];
                item = item.Trim();
                if (!string.IsNullOrEmpty(item))
                    newLine.Add(item);
            }
            newLine = newLine.Distinct().ToList();
            line = string.Join(", ", newLine);
            lines[i] = line;
        }

        File.WriteAllLines(fileName, lines);
    }
}

File: tagbag/double_tag_deleter.py
# ---------------------------------
# File processors
# ---------------------------------

from .tag_base import TagBase
from .tag_helper import TagHelper


from pathlib import Path
from typing import List, Sequence, Set


class DoubleTagDeleter(TagBase):
    """
    Remove duplicate tags within each line of every file.
    """

    @staticmethod
    def execute(args: Sequence[str]) -> None:
        """
        args[0]: directory path
        """
        file_names = TagBase.get_tag_files(args[0])
        for name in file_names:
            DoubleTagDeleter._process_file(Path(name))

    @staticmethod
    def _process_file(file_path: Path) -> None:
        lines = file_path.read_text(encoding="utf-8").splitlines()
        new_lines: List[str] = []

        for line in lines:
            items = [t.strip() for t in TagHelper.split_into_tags(line) if t.strip()]
            seen: Set[str] = set()
            deduped = [t for t in items if not (t in seen or seen.add(t))]
            new_lines.append(", ".join(deduped))

        file_path.write_text("\n".join(new_lines), encoding="utf-8")